spylon/spark/yarn_launcher.py

```diff
@@ -258,84 +313,53 @@ def launcher(deploy_mode, args, working_dir=".", cleanup=True):
     elif conda_env.endswith(".zip"):
-        log.info("Using conda environment from zip file")
-        env_archive = conda_env
-        basename, _ = os.path.splitext(env_archive)
-        env_name = os.path.basename(basename)
-        env_dir = _extract_local_archive(working_dir, cleanup_functions, env_name, env_archive)
-
-    # The case where we have to CREATE the environment ourselves
+        # We have a precreated conda environment around.
+        env_name, env_dir, conda_env = _conda_from_zip(**func_kwargs)
     elif conda_env.endswith(".yaml"):
-        log.info("Building conda environment from yaml specification")
-        with open(conda_env) as fo:
-            env = yaml.load(fo)
-
-        conda_create_args = env.get("conda-args", "")
-        conda_create_args = shlex.split(conda_create_args)
-        deps = env["dependencies"]
-        env_name = env.get("name", "envname")
-
-        # Create the conda environment
-        env_dir, env_name = create_conda_env(working_dir, env_name, deps, conda_create_args)
-
-        # Archive the conda environment
-        env_archive = archive_dir(env_dir)
-
-        abs_env_dir = os.path.abspath(env_dir)
-        abs_env_archive = os.path.abspath(env_archive)
-
-        cleanup_functions.extend([
-            lambda: shutil.rmtree(abs_env_dir),
-            lambda: os.unlink(abs_env_archive),
-        ])
+        # The case where we have to CREATE the environment ourselves
+        env_name, env_dir, env_archive = _conda_from_yaml(**func_kwargs)
     else:
-        raise NotImplementedError()
-
-    args = dict(env_dir=env_dir, env_name=env_name, env_archive=env_archive, args=spark_args)
+        raise NotImplementedError("Can only run launcher if your conda env is on hdfs (starts "
+                                  "with 'hdfs:/', is already a zip (ends with '.zip'), or is "
+                                  "coming from a yaml specification (ends with '.yaml' and "
+                                  "conforms to the conda environment.yaml spec)")
+    del func_kwargs
+
+    func_kwargs = dict(env_dir=env_dir, env_name=env_name, env_archive=env_archive, args=spark_args)
+    funcs = {'client': run_pyspark_yarn_client, 'cluster': run_pyspark_yarn_cluster}
     try:
-        if deploy_mode == "client":
-            run_pyspark_yarn_client(**args)
-        elif deploy_mode == "cluster":
-            run_pyspark_yarn_cluster(**args)
+        funcs[deploy_mode](**func_kwargs)
     finally:
-        if cleanup:
-            for fn in cleanup_functions:
-                try:
-                    fn()
-                except:
-                    import traceback
-                    traceback.print_exc()
+        if not cleanup:
+            return
+        # iterate over and call all cleanup functions
+        for function in cleanup_functions:
+            try:
+                function()
+            except:
+                # Why are we catching this and just printing the traceback?
```

Comment:
I'm guessing so that all the cleanup functions get to run, even if any one of them fails. This module has a logger. `log.exception("Cleanup function %s failed", function)` would be a better approach since `exception` prints the traceback and let's us include a short log message.